{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import rank_mod\n",
    "import build\n",
    "import argparse\n",
    "import sys\n",
    "import iohelp\n",
    "import variant\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pysam import FastaFile\n",
    "import resource\n",
    "import time\n",
    "from util import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "varstest = '../test/sm.1ksnp'\n",
    "referencetest = '../test/sm.fa'\n",
    "chromtest = '22'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "method = 'popcov-blowup'\n",
    "reference = referencetest #'../../hs37d5.fa'\n",
    "vars = varstest #'../../chr6.1ksnp'\n",
    "window_size = 100\n",
    "output1 = '../../'+method+'-mod.txt'\n",
    "chrom = chromtest#'6'\n",
    "rankpath = output1\n",
    "pct = 10\n",
    "output2 = '../../build-mod-'+str(pct/100)+'.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = argparse.ArgumentParser(description=__doc__,\n",
    "            formatter_class=argparse.RawDescriptionHelpFormatter)\n",
    "parser.add_argument('--method', default=method)\n",
    "parser.add_argument('--reference', default=reference)\n",
    "parser.add_argument('--vars', default=vars)\n",
    "parser.add_argument('--window-size', default=window_size)\n",
    "parser.add_argument('--chrom', default=chrom)\n",
    "parser.add_argument('--output', default=output1)\n",
    "\n",
    "parser.add_argument('--prune', type=int, required=False)\n",
    "parser.add_argument('--phasing', type=str, required=False)\n",
    "parser.add_argument('--pseudocontigs', action=\"store_true\")\n",
    "\n",
    "parser.add_argument('--hisat', type=str, required=False, help='Path to file to write HISAT2 --snp information')\n",
    "parser.add_argument('--erg', type=str, required=False, help='Path to fasta file to write additional pseudocontigs for ERG alignment')\n",
    "parser.add_argument('--sorted', type=str, default=rankpath)\n",
    "parser.add_argument('--pct', type=int, required=False, help='Percentage of variants to include in graph genome')\n",
    "\n",
    "\n",
    "args = parser.parse_args(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "parser1 = argparse.ArgumentParser(description=__doc__,\n",
    "            formatter_class=argparse.RawDescriptionHelpFormatter)\n",
    "parser1.add_argument('--method', default=method)\n",
    "parser1.add_argument('--reference', default=reference)\n",
    "parser1.add_argument('--vars', default=vars)\n",
    "parser1.add_argument('--window-size', default=window_size)\n",
    "parser1.add_argument('--chrom', default=chrom)\n",
    "parser1.add_argument('--output', default=output2)\n",
    "\n",
    "parser1.add_argument('--prune', type=int, required=False)\n",
    "parser1.add_argument('--phasing', type=str, required=False)\n",
    "parser1.add_argument('--pseudocontigs', action=\"store_true\")\n",
    "\n",
    "parser1.add_argument('--hisat', type=str, required=False, help='Path to file to write HISAT2 --snp information')\n",
    "parser1.add_argument('--erg', type=str, required=False, help='Path to fasta file to write additional pseudocontigs for ERG alignment')\n",
    "parser1.add_argument('--sorted', type=str, default=rankpath)\n",
    "parser1.add_argument('--pct', type=int, default=pct, required=False, help='Percentage of variants to include in graph genome')\n",
    "\n",
    "\n",
    "args1 = parser1.parse_args(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def go(args):\n",
    "    start_time = time.time()\n",
    "    x1 = 'Memory usage start: %s (kb)' % resource.getrusage(resource.RUSAGE_SELF).ru_maxrss\n",
    "    if args.window_size:\n",
    "        r = args.window_size\n",
    "    else:\n",
    "        r = 35\n",
    "    if args.output:\n",
    "        o = args.output\n",
    "    else:\n",
    "        o = 'ordered.txt'\n",
    "    if args.prune:\n",
    "        max_v = args.prune\n",
    "    else:\n",
    "        max_v = r\n",
    "    genome1 = {args.chrom: ''}\n",
    "    if args.method == 'hybrid':\n",
    "        genome1 = iohelp.read_genome1(args.reference, args.chrom)\n",
    "    x2 = 'Memory usage after  genome created: %s (kb)' % resource.getrusage(resource.RUSAGE_SELF).ru_maxrss\n",
    "    vars1 = iohelp.parse_1ksnp1(args.vars)\n",
    "    x3 = 'Memory usage after  variants created: %s (kb)' % resource.getrusage(resource.RUSAGE_SELF).ru_maxrss\n",
    "    ranker = VarRanker1(genome1, vars1, r, args.phasing, max_v)\n",
    "    ranked = ranker.rank(args.method, o)\n",
    "    x4 = 'Memory usage after  ranking created: %s (kb)' % resource.getrusage(resource.RUSAGE_SELF).ru_maxrss\n",
    "    x5 = \"Running time: %s seconds\" % (time.time() - start_time)\n",
    "    oram = 'ram-'+args.method+'-mod.txt'\n",
    "    with open(oram, 'w') as f:\n",
    "        f.writelines([x1,'\\n',x2,'\\n',x3,'\\n',x4,'\\n',x5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "class VarRanker1:\n",
    "    def __init__(self, genome, variants, r, phasing, max_v):\n",
    "        self.genome = genome\n",
    "        self.chrom_lens = dict()\n",
    "        for chrom, seq in genome.items():\n",
    "            self.chrom_lens[chrom] = len(seq)\n",
    "\n",
    "        self.variants = variants\n",
    "        self.num_v = self.variants.shape[0]\n",
    "        self.r = r\n",
    "\n",
    "        self.phasing = phasing\n",
    "        self.hap_parser = iohelp.HaplotypeParser(phasing) if phasing else None\n",
    "\n",
    "        self.max_v_in_window = max_v\n",
    "\n",
    "        self.h_ref = None\n",
    "        self.h_added = None\n",
    "\n",
    "        self.wgt_ref = None\n",
    "        self.wgt_added = None\n",
    "\n",
    "        self.curr_vars = None\n",
    "        self.freqs = {}\n",
    "\n",
    "    def rank(self, method, out_file):\n",
    "        ordered = None\n",
    "        ordered_blowup = None\n",
    "        print(method)\n",
    "        if method == 'popcov':\n",
    "            ordered = self.rank_pop_cov()\n",
    "        elif method == 'popcov-blowup':\n",
    "            ordered = self.rank_pop_cov(True)\n",
    "        elif method == 'hybrid':\n",
    "            ordered, ordered_blowup = self.rank_hybrid()\n",
    "\n",
    "        if ordered is not None:\n",
    "            df = pd.DataFrame(columns = ['a', 'b'])\n",
    "            df['a'] = self.variants.loc[ordered].chrom\n",
    "            df['b'] = self.variants.loc[ordered].pos +1\n",
    "            df.to_csv(out_file,header=False,index=False,line_terminator = '\\t')\n",
    "            # with open(out_file, 'w') as f:\n",
    "            #     f.write('\\t'.join([str(self.variants.chrom[i]) + ',' for i in ordered]))\n",
    "        if ordered_blowup:\n",
    "            df = pd.DataFrame(columns = ['a', 'b'])\n",
    "            df['a'] = self.variants.loc[ordered_blowup].chrom\n",
    "            df['b'] = self.variants.loc[ordered_blowup].pos +1\n",
    "            df.to_csv(out_file,header=False,index=False,line_terminator = '\\t')\n",
    "            # with open(out_file+'.blowup', 'w') as f:\n",
    "            #     f.write('\\t'.join([self.variants.chrom[i] + ',' + str(self.variants.pos[i]+1) for i in ordered_blowup]))\n",
    "\n",
    "    def rank_pop_cov(self, with_blowup=False, threshold=1/3):\n",
    "        '''\n",
    "        Rank variants using the hybrid ranking method.\n",
    "        \n",
    "        with_blowup: If true, add blowup penalty for neighboring variants\n",
    "        threshold: Blowup penalty for neighboring variants, between 0 and 1. A smaller value is a stricter penalty\n",
    "        '''\n",
    "        \n",
    "        if with_blowup:\n",
    "            varsdf = self.variants\n",
    "            var_wgts = (self.variants.iloc[:,[4,8,9]].sum(axis = 1)).sort_index()\n",
    "            posdf = varsdf.pos\n",
    "            ffirst = np.vectorize(\n",
    "                lambda i: \n",
    "                    (posdf <= posdf[i]-r).idxmin()\n",
    "                    )\n",
    "\n",
    "            flast = np.vectorize(\n",
    "                lambda i: \n",
    "                    (posdf >= posdf[i]+r).idxmax()\n",
    "                    )\n",
    "\n",
    "            first = np.maximum(0,np.fromfunction(ffirst,(self.num_v,), dtype=int))\n",
    "            last = np.minimum(self.num_v,np.fromfunction(flast,(self.num_v,), dtype=int))-1\n",
    "            last[last == -1] = self.num_v-1\n",
    "            neighbors = last-first\n",
    "            penalty = np.repeat(threshold,self.num_v)\n",
    "            var_wgts = pd.Series(var_wgts* (penalty**neighbors).copy()).sort_values(ascending=False)\n",
    "            print(var_wgts)\n",
    "            ordered = var_wgts.index\n",
    "        else:\n",
    "            # Variant weight is the sum of frequencies of alternate alleles\n",
    "            var_wgts = (self.variants.iloc[:,[4,8,9]].sum(axis = 1)).sort_index()\n",
    "            var_wgts = var_wgts.sort_values(ascending=False)\n",
    "            ordered = var_wgts.index\n",
    "\n",
    "        return ordered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_genome1(filename, target_chrom=None):\n",
    "    rec = FastaFile(filename)\n",
    "    lengths = {rec.references[i]:rec.lengths[i] for i in range(len(rec.lengths))}\n",
    "    return( {target_chrom:rec.fetch(reference=target_chrom, start = 0, end = lengths[target_chrom])})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_1ksnp1(filename, G=None):\n",
    "    df = pd.read_csv(filename, sep = \"\\t\",header=None)\n",
    "    df.columns = [\"chrom\", \"pos\", \"orig\", \"alts\", \"probs1\", 'x', 'num', \"name\"]\n",
    "    df.pos = df.pos-1\n",
    "    df['probs2'] = np.zeros((df.shape[0]))\n",
    "    df['probs3'] = np.zeros((df.shape[0]))\n",
    "    rem = []\n",
    "    for num in range(1,4):\n",
    "        indices = pd.Series(df.index[df.num == num])\n",
    "        s = indices.iloc[np.arange(len(indices)//num)*num]\n",
    "        for j in range(1,num):\n",
    "            col = 'probs'+str(j+1)\n",
    "            df[col].iloc[s] = np.array(df['probs1'].iloc[s+j])\n",
    "            df['alts'].iloc[s] += np.array(df['alts'].iloc[s+j])\n",
    "            rem.append(s+j)\n",
    "    df = df.drop(pd.concat(rem).sort_values()).reset_index(drop=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "popcov-blowup\n",
      "asdas\n",
      "genome check\n",
      "variants check\n",
      "[('22', 11190), ('22', 161571), ('22', 8010), ('22', 294874), ('22', 102546), ('22', 402330), ('22', 93886), ('22', 109000), ('22', 295341), ('22', 100529)]\n",
      "Found 41 / 41 variants\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/martin/Documents/VSCode/Proyecto BCOM4006/FORGe-mod/src/iohelp.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[col].iloc[s] = np.array(df['probs1'].iloc[s+j])\n",
      "/home/martin/Documents/VSCode/Proyecto BCOM4006/FORGe-mod/src/iohelp.py:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['alts'].iloc[s] += np.array(df['alts'].iloc[s+j])\n"
     ]
    }
   ],
   "source": [
    "rank_mod.go(args)\n",
    "print(\"asdas\")\n",
    "build.go(args1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.6 ('biopython')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "f95aa36fe6ff1928aaa9a1cef51ad30cd5fd6bf1f8fec5b5f7edbc2658fcc0f9"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
